skaffold.yaml

```diff
@@ -69,13 +87,31 @@ manifests:
     paths:
     - local-setup
 deploy:
-  kubectl: {}
+  # --server-side apply is a workaround for https://github.com/gardener/gardener/issues/10267.
+  # kubectl apply attempts a strategic merge patch which fails for a ControllerDeployment.
+  # For more details, see https://github.com/gardener/gardener/issues/10267.
+  #
+  # TODO: Switch back to "kubectl: {}" when the above issue is resolved.
+  kubectl:
+    flags:
+      apply:
+        - --server-side
+        - --force-conflicts
 ---
-apiVersion: skaffold/v4beta3
+apiVersion: skaffold/v4beta7
 kind: Config
 metadata:
   name: admission
 build:
+  tagPolicy:
+    customTemplate:
+      template: "{{.version}}-{{.sha}}"
+      components:
+        - name: version
+          envTemplate:
+            template: "{{.EXTENSION_VERSION}}"
+        - name: sha
+          inputDigest: {}
```

Comment:
I am not sure that it will work with your suggestion. I haven't tried and I won't have capacity this week to try it out locally. 
I am not sure if https://github.com/gardener/gardener/blob/01391f78c30a37968fceafd885f0637e689f1909/skaffold.yaml#L1696-L1706 will help in the case you have changes locally which are not committed. With inputDigest I want to cover the case where you change the code (without local commit) and skaffold generates a new tag.

Reply:
Yep, having changes locally which are not committed is covered. From the [tagging docs](https://skaffold.dev/docs/taggers/#how-tagging-works)
> When images are pushed, their immutable digest is available. Skaffold then references images both by tag and digest. Something like image:tag@sha256:abacabac.... Using both the tag and the digest seems superfluous but it guarantees immutability and helps users quickly see which version of the image is used.

I also tested it locally. Without any file changes I got the following image for the `admission` deployment:

> image: garden.local.gardener.cloud:5001/local-skaffold_gardener-extension-shoot-rsyslog-relp-admission:v0.7.0-dev-6b6e603-dirty@sha256:f68ae005709b2b217d02d75f50612f2c39c8bf5250909a30f9867f395e591f39

After adding changes to a file, part of the admission, and executing `make extensions-up` (without committing the changes in git), I got an image with the same tag, but a different sha digest:

> image: garden.local.gardener.cloud:5001/local-skaffold_gardener-extension-shoot-rsyslog-relp-admission:v0.7.0-dev-6b6e603-dirty@sha256:852ce3df6d2c7f91b861f42edcefb1f24651e851eb72eef2bcb103f31cf37581

This is how it used to work before. The actual reason for the issue was that we switched to using https://github.com/gardener/gardener-extension-shoot-rsyslog-relp/blob/main/hack/generate-kustomize-patch-controllerdeployment-shoot-rsyslog-relp.sh to generate the `ControllerDeployment` as the `SKAFFOLD_IMAGE` env variable does not contain the sha digest that would differ depending on local code changes (it only contains the tag). That's why we need to use `inputDigest` when building the `extension` image for the `ControllerDeployment` to manually add this digest to the tag.

Anyway, I was initially confused that there were no problems some time ago when making local code changes, and by `skaffold` in gardener where `inputDigest` was only used for `provider-local`. Adding a doc to the tag for the `extension` config [here](https://github.com/gardener/gardener-extension-shoot-rsyslog-relp/blob/33b326500d50512447a01a4d9a34ec3f352b2cfa/skaffold.yaml#L46-L47) to explain this would be useful. Maybe something in the lines of
```
// inputDigest is used to inject a digest of the artifact source into the built image tag and therefore into the `SKAFFOLD_IMAGE` environment variable which is used when generating the corresponding `ControllerDeployment`
```

Reply:
I pushed https://github.com/gardener/gardener-extension-shoot-rsyslog-relp/pull/194/commits/3816113ff8a3344e5bfd29dd8d0cb849e9bbca6b to address your suggestion.

Reply:
FYI @rrhubenov as you wanted to the the same changes to the lakom extension.